### ISSUE TITLE
Bug: unhandled error in `lintFiles` with glob patterns when config is invalid

### ISSUE THREAD

**Issue by fasttime** [eslint/eslint#18250]
### Environment

Node version: v21.7.1
npm version: v10.5.0
Local ESLint version: v9.0.0-rc.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 23.4.0

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```
export default {
  invalid: 'foobar',
};
```
</details>

```js
// index.js
import { ESLint } from 'eslint';

const eslint = new ESLint();
try {
  await eslint.lintFiles('*.js');
} catch {
  console.log('An error should be caught because the config is invalid');
}
console.log('This message should be printed');
```

### What did you expect to happen?

`lintFiles` should throw an error in the current execution context because the config is invalid.

### What actually happened?

The process terminated abnormally while running `lintFiles` because of an uncaught error in an asynchronous callback.

```console
$ node ./index.js
.../project/node_modules/@humanwhocodes/object-schema/src/object-schema.js:270
                throw new UnexpectedKeyError(key);
                ^

UnexpectedKeyError: Unexpected key "invalid" found.
    at ObjectSchema.validate (.../project/node_modules/@humanwhocodes/object-schema/src/object-schema.js:270:23)
    at .../project/node_modules/@humanwhocodes/object-schema/src/object-schema.js:239:18
    at Array.reduce (<anonymous>)
    at ObjectSchema.merge (.../project/node_modules/@humanwhocodes/object-schema/src/object-schema.js:237:24)
    at .../project/node_modules/@humanwhocodes/config-array/api.js:935:42
    at Array.reduce (<anonymous>)
    at FlatConfigArray.getConfig (.../project/node_modules/@humanwhocodes/config-array/api.js:934:39)
    at FlatConfigArray.isFileIgnored (.../project/node_modules/@humanwhocodes/config-array/api.js:962:15)
    at .../project/node_modules/eslint/lib/eslint/eslint-helpers.js:322:49
    at Array.reduce (<anonymous>)
```

The callback in question is the [`entryFilter` of `@nodelib/fs.walk`](https://github.com/eslint/eslint/blob/de408743b5c3fc25ebd7ef5fb11ab49ab4d06c36/lib/eslint/eslint-helpers.js#L291-L331), which apparently is not expected to throw errors.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/stackblitz-starters-swoxsq

### Participation

- [X] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
I can reproduce this. Any ideas how to fix it?

**Comment by fasttime** [comment]
> Any ideas how to fix it?

We could catch the error in `entryFilter` and rethrow it after [`doFsWalk`](https://github.com/eslint/eslint/blob/de408743b5c3fc25ebd7ef5fb11ab49ab4d06c36/lib/eslint/eslint-helpers.js#L283-L333). Or we could call `fswalk.walk` inside a promise executor (instead of wrapping it with `util.promisify`), so we could reject the promise when `entryFilter` throws an error.

Finally, we could open an issue at [`nodelib`](https://github.com/nodelib/nodelib) and ask them to handle errors in their filter callbacks, so we don't have to. Upgrading to `@nodelib/fs.walk` 2.0.0 didn't help, and I couldn't find a closed issue in their repo that matches our problem.

**Comment by mdjermanovic** [comment]
> Finally, we could open an issue at [`nodelib`](https://github.com/nodelib/nodelib) and ask them to handle errors in their filter callbacks, so we don't have to. Upgrading to `@nodelib/fs.walk` 2.0.0 didn't help, and I couldn't find a closed issue in their repo that matches our problem.

This seems like the best solution, but it will probably take some time. Can you open an issue there?

> Or we could call `fswalk.walk` inside a promise executor (instead of wrapping it with `util.promisify`), so we could reject the promise when `entryFilter` throws an error.

This would be the easiest solution for us at the moment, though I wouldn't expect it to work if `util.promisify` doesn't, but let's try it.

**Comment by fasttime** [comment]
Thanks for the feedback @mdjermanovic.

> > Finally, we could open an issue at [`nodelib`](https://github.com/nodelib/nodelib) and ask them to handle errors in their filter callbacks, so we don't have to. Upgrading to `@nodelib/fs.walk` 2.0.0 didn't help, and I couldn't find a closed issue in their repo that matches our problem.
> 
> This seems like the best solution, but it will probably take some time. Can you open an issue there?

Sure!

> > Or we could call `fswalk.walk` inside a promise executor (instead of wrapping it with `util.promisify`), so we could reject the promise when `entryFilter` throws an error.
> 
> This would be the easiest solution for us at the moment, though I wouldn't expect it to work if `util.promisify` doesn't, but let's try it.

I can look into this and see what work works for us.

**Comment by nzakas** [comment]
Issue: https://github.com/nodelib/nodelib/issues/105

**Comment by pumano** [comment]
got same error when using nx and config contains:
```js
 languageSettings: { parserOptions: { project: ["./tsconfig.*?.json"] } }
 ```
 
looks like schema does not contain that field and error handling processed incorrectly (as mentioned above)

console output:
```bash
node_modules/@nx/eslint/node_modules/@humanwhocodes/object-schema/src/object-schema.js:270
                      throw new UnexpectedKeyError(key);
                      ^
      
      UnexpectedKeyError: Unexpected key "languageSettings" found.

```
